File: [2] Calculations/Pizza-Party/PizzaParty.py
# Write a program to evenly divide pizzas.
# Prompt for the number of people, the number of pizzas, and the number of slices per pizza.
# Ensure that the number of pieces comes out even.
# Display the number of pieces of pizza each person should get.
# If there are leftovers, show the number of leftover pieces.
import math


# Revise the program to ensure that inputs are entered as numeric values.
# Don't allow the user to proceed if the value entered is not numeric.
def ValidatedNumber(prompt, _min=None, _max=None):
    while True:
        user_input = input(prompt)
        try:
            number = float(user_input)
            if (_min is None or _min <= number) and (_max is None or _max >= number):
                return number
            else:
                print(f"Input out of range. [{_min}-{_max}]")
        except ValueError:
            print("! Error: Input is NaN")


class PizzaSplitter:
    def __init__(self):
        self.people = ValidatedNumber("How many people? ")
        self.pizzas = ValidatedNumber("How many pizzas do you have? ")
        self.slices = ValidatedNumber("How many slices per pizza? ") * self.pizzas
        self.pieces = int(self.slices / self.people)
        self.leftovers = self.slices % self.people

        # Alter the output so it handles pluralization properly.
        print(f"{int(self.people)} people with {int(self.pizzas)} pizzas")
        if self.pieces == 1:
            print(f"Each person gets 1 piece of pizza.")
        else:
            print(f"Each person gets {int(self.pieces)} pieces of pizza.")

        # Handle the output for leftover pieces appropriately as well.
        if self.leftovers == 0:
            print("There are no leftovers.")
        else:
            print(f"There are {int(self.leftovers)} leftover pieces.")


# Create a variant of the program that prompts for the number of people
# and the number of pieces each person wants,
# and calculate how many full pizzas you need to purchase.
class PizzaBuyer:
    def __init__(self):
        self.people = ValidatedNumber("How many people? ")
        self.slices = ValidatedNumber("How many slices are there per pizza? ")
        self.pieces = ValidatedNumber("How many slices does each person want? ")
        total_slices = self.people * self.pieces
        self.pizzas = math.ceil(total_slices / self.slices)
        self.leftovers = (self.pizzas * self.slices) - total_slices

        # Alter the output so it handles pluralization properly.
        print(f"{int(self.people)} people with {int(self.pieces)} slices each")
        if self.pizzas <= 1:
            print(f"You need to buy 1 full pizza.")
        else:
            print(f"You need to buy {int(self.pizzas)} full pizzas.")

        # Handle the output for leftover pieces appropriately as well.
        if self.leftovers == 0:
            print("There will be no leftovers.")
        else:
            print(f"There will be {int(self.leftovers)} leftover pieces.")


print("[1] I want to know how many slices each person should get.")
print("[2] I want to know how many pizzas I have to buy.")
choice = ValidatedNumber("Choose an option: ", 1, 2)
party = choice == 1 and PizzaSplitter() or PizzaBuyer()

File: [2] Calculations/Pizza-Party/PizzaParty.cs
using System;
using static CSharp._2__Calculations.DataValidation;

namespace CSharp._2__Calculations.Pizza_Party
{
    internal class PizzaParty
    {
        static class Pizza
        {
            private static short people;
            private static short pizzas;
            private static short slices;
            private static short pieces;
            private static short leftovers;
            public static void Splitter()
            {
                // Input
                people = (short)ValidatedNumber("How many people? ");
                pizzas = (short)ValidatedNumber("How many pizzas do you have? ");
                slices = (short)(ValidatedNumber("How many slices per pizza? ") * pizzas);

                // Process
                pieces = (short)(slices / people);
                leftovers = (short)(slices % people);

                // Output
                Console.WriteLine($"{people} people with {pizzas} pizzas");
                if (pieces == 1)
                    Console.WriteLine($"Each person gets 1 piece of pizza.");
                else
                    Console.WriteLine($"Each person gets {pieces} pieces of pizza.");

                if (leftovers == 0)
                    Console.WriteLine("There are no leftovers.");
                else
                    Console.WriteLine($"There are {leftovers} leftover pieces.");
            }
            public static void Buyer()
            {
                // Input
                people = (short)ValidatedNumber("How many people? ");
                slices = (short)ValidatedNumber("How many slices are there per pizza? ");
                pieces = (short)(ValidatedNumber("How many slices does each person want? "));

                // Process
                var total_slices = (short)(people * pieces);
                pizzas = (short)Math.Ceiling((decimal)(total_slices / slices));
                leftovers = (short)(slices % people);

                // Output
                Console.WriteLine($"{people} people with {pieces} slices each");
                if (pieces <= 1)
                    Console.WriteLine($"You need to buy 1 full pizza.");
                else
                    Console.WriteLine($"You need to buy {pizzas} full pizzas.");

                if (leftovers == 0)
                    Console.WriteLine("There will be no leftovers.");
                else
                    Console.WriteLine($"There will be {leftovers} leftover pieces.");
            }
        }

        static void Main()
        {
            Console.WriteLine("[1] I want to know how many slices each person should get.");
            Console.WriteLine("[2] I want to know how many pizzas I have to buy.");
            short choice = (short)ValidatedNumber("Choose an option: ", _min: 1, _max: 2, _type: "short");
            if (choice == 1)
                Pizza.Splitter();
            else
                Pizza.Buyer();
        }
    }
}
